net: ip: net_pkt: set pkt context, iface and family in net_pkt_get()

mike-scott opened the pull request:
Commit 753daa6 ("net: pkt: Compute TX payload data length")
removed the default packet setup on incoming packets when they
belong to the rx_pkt pool.

Let's restore this behavior, as MBEDTLS processing in net_app library
needs to use packet family to determine IP header length on
incoming packets.

Also, due to the way MTU is checked in net_pkt_append, let's set the
packet's data_len field to a nice high value so that it doesn't get
in the way.

NOTE: A future cleanup patch could set the IP header length based
on the context IP family.  However, there are many places in the code
where this is being set, so care should be taken.

Signed-off-by: Michael Scott <michael@opensourcefoundries.com>

mike-scott commented:
NOTE: This fixes DTLS on LwM2M and is required for: https://github.com/zephyrproject-rtos/zephyr/pull/5267

pfalcon commented:
@pfl : FYI.

rbtchc reviewed (approved):
LGTM

pfl commented on subsys/net/lib/app/net_app.c:
```diff
@@ -2066,6 +2066,9 @@ int _net_app_ssl_mainloop(struct net_app_ctx *ctx)
 				hdr_len = ctx->tls.mbedtls.ssl_ctx.hdr->len;
+				/* set packet family for checks below */
+				net_pkt_set_family(pkt,
+					net_context_get_family(net_ctx));
```
Shouldn't this setup be part of  subsys/net/ip/net_pkt.c instead? It got removed from there with commit #753daa6, but only because it wasn't needed by those changes. At the time this usage wasn't caught, so perhaps there may be more code with the same semantics.

Now that commit #753daa6 and later have correctly computed pkt->data_len, couldn't that information be used here instead of re-computing the data length? Or is this only about IP header lengths and not net_pkt payload length?

mike-scott replied:
@pfl You are probably correct.  Along those lines, I can re-implement code in net_pkt.c net_pkt_get() to be executed regardless of whether slab != &rx_pkts to do the following:
- net_pkt_set_context
- net_pkt_set_iface
- net_pkt_set_family

As these could probably be used elsewhere in the code regardless of the slab used.

To clarify: the checks are used for header lengths, not payload data length, so the data_len calc doesn't help.

lpereira replied:
If you can get by with getting the address family for the checks below (under `#ifdefs`), why not just call `net_context_get_family(net_ctx)` instead or `net_pkt_family(pkt)`?  Otherwise, yeah, @pfl's suggestion makes sense.

mike-scott replied:
@lpereira I can get by with the using net_context_get_family in the checks below, however, I think @pfl was wondering if other code may have been affected by the commit removing the initialization of several packet values based on the slap used.

mike-scott replied:
@pfl can you let me know if my understanding of your previous statements is correct above?

pfl replied:
Yes it is. But now that I look again on the commit/source code, it does set all the details, but only when slab != &rx_pkts. There was a reason for this, but @jukkar is away right now so I can't re-iterate on the why... So you end up using the rx_pkts slab in your code?

rveerama1 replied:
[net_core.c](https://github.com/rveerama1/zephyr/blob/master/subsys/net/ip/net_core.c#L108) will start setting (family, proto) on incoming(rx) packets. Then UDP handler handover the packet to net_context connection ([packet_received](https://github.com/rveerama1/zephyr/blob/master/subsys/net/ip/net_context.c#L2230)). But in [net_app with tls](https://github.com/rveerama1/zephyr/blob/master/subsys/net/lib/app/net_app.c#L2056) creates **new** packet for decrypted payload. This is a fresh new packet which doesn't have any information (iface, family, context. etc..) because of slab != &rx_pkts. @pfl MTU handling required for only Tx packets and all setting calls slipped to if condition which might be the cause.
I would suggest net_pkt_set_context, net_pkt_set_iface, net_pkt_set_family, net_pkt_set_ip_hdr_len could be set irrespective of "slab != &rx_pkts" condition.

pfl replied:
> I would suggest net_pkt_set_context, net_pkt_set_iface, net_pkt_set_family, net_pkt_set_ip_hdr_len could be set irrespective of "slab != &rx_pkts" condition.

That seems to be the change we are looking for.

mike-scott replied:
Today, I noticed that I saw getting a hidden error in net_pkt:
[net/conn] [DBG] net_conn_input: (0x00403500): [0] match found cb 0x0000ed83 ud 0x00400660 rank 0x33
*[net/pkt] [ERR] net_frag_read: Not enough data to read*
*[net/udp] [ERR] net_udp_get_hdr: {assert: 'frag' failed}*
[lib/lwm2m_engine] [DBG] lwm2m_udp_receive: checking for reply from [2001:db8::2]

This doesn't seem to affect the incoming data when I use this patch as the registration works, but when I re-add the default packet setup in net_pkt.c it doesn't work at all.

Still taking a look at this.

mike-scott committed:
net: ip: net_pkt: set pkt context, iface and family in net_pkt_get()

Commit 753daa6 ("net: pkt: Compute TX payload data length")
removed the default packet setup on incoming packets when they
belong to the rx_pkt pool.

Let's restore this behavior, as MBEDTLS processing in net_app library
needs to use packet family to determine IP header length on
incoming packets.

NOTE: A future cleanup patch could set the IP header length based
on the context IP family.  However, there are many places in the code
where this is being set, so care should be taken.

Signed-off-by: Michael Scott <michael@opensourcefoundries.com>

mike-scott commented:
@pfl Can you take a look at the new change to this patch?
Before the MTU changes here: https://github.com/zephyrproject-rtos/zephyr/commit/753daa65f9181b8ea89a1597b7aaa30a45910f44#diff-36fffa10c9a7f7bf15b267b17920b155

net_pkt_append() had it's own calculation for how much data could be appended to the buffer when not using TCP:
			struct net_if *iface = net_context_get_iface(ctx);
			max_len = net_if_get_mtu(iface);
			/* Optimize for number of jumps in the code ("if"
			 * instead of "if/else").
			 */
			max_len -= NET_IPV4TCPH_LEN;
			if (net_context_get_family(ctx) != AF_INET) {
				max_len -= NET_IPV6TCPH_LEN - NET_IPV4TCPH_LEN;
			}

How would you like me to handle this now?  We're using the pkt->data_len value (which is currently not set for incoming packets).

Currently, I dummied in 0x1000 to confirm that this was causing the issues I was seeing with this patch.

mike-scott commented:
@pfl after looking through the older code and comparing to the newer code, I have to ask why we're filtering on "if (pkt && slab != &rx_pkts) {"?

What about the MTU code doesn't apply to handling rx packets?

pfl commented:
> How would you like me to handle this now? We're using the pkt->data_len value (which is currently not set for incoming packets).

For incoming packets the "MTU" or actually packet length in that case is known only on packet reception. When acquiring a net_pkt or similar to hold the RX data, we don't have a clue yet, as the incoming packet length is not known. @jukkar mentioned that the RX path will figure it out. It would of course be nice to fill in this value to pkt->data_len, but it's not really needed as there is nothing we can do with or for the MTU except accept the fact that a finite amount of bytes were sent by the sender.

pfl commented on subsys/net/ip/net_pkt.c:
```diff
@@ -564,6 +567,12 @@ static struct net_pkt *net_pkt_get(struct k_mem_slab *slab,
+		 * In case, net_pkt_append() is later used on this packet,
+		 * set the RX packet data_len to a nice high number
+		 */
+		pkt->data_len = 0x1000;
```
This we can't set, as we don't know the lenght of the incoming packet. pkt->data_len needs not be consulted on receiving, only when we attempt to send.

mike-scott replied:
The problem is that the DTLS receive path for net_app uses net_pkt_append() and it requires data_len be set if the context is set: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/ip/net_pkt.c#L1220

Otherwise you get max_len 0.

mike-scott replied:
FYI: Here's the DTLS receive path in net_app where net_pkt_append is used: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/app/net_app.c#L2087

Summary:
- Allocate the packet via net_pkt_get_rx() which calls net_pkt_get()
- Add IP UDP/TCP headers if found
- Append the decrypted data from request_buf via net_pkt_append_all() which calls net_pkt_append()

We would need the pkt->data_len to be set somewhere before the data append for it to work correctly.  Otherwise we can't set the net_context of the packet.

pfl commented on subsys/net/ip/net_pkt.c:
```diff
@@ -520,19 +521,21 @@ static struct net_pkt *net_pkt_get(struct k_mem_slab *slab,
-
 		if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
 			data_len = max(iface_len, NET_IPV6_MTU);
 			data_len -= NET_IPV6H_LEN;
```
The above looks fine (as long as test cases also agree that it works).

mike-scott committed:
net: net_app: set packet data_len on incoming packets

This fixes a bug when calling net_pkt_append[_all] which uses
pkt->data_len as part of the maximum packet length calculation
when the net_context is set.

Without this change the maximum packet length is calculated as 0
(the value of pkt->data_len) and an ENOMEM error is returned.

Signed-off-by: Michael Scott <michael@opensourcefoundries.com>

mike-scott commented:
New version of this patch posted (and rebased on today's master):
I split the behavior into 2 sections:
1. Re-add the removed portion of the default packet setup setting net_context, iface and family in net_pkt.c
2. Use the incoming request_buf len to set the pkt->data_len appropriately so that net_pkt_append can be called w/o error.

pfalcon commented:
@mike-scott, First of all, sorry for the delay with looking into this.

> We're using the pkt->data_len value (which is currently not set for incoming packets).

Can you please elaborate on that?

And given that you're likely on holidays (Merry Xmas/Happy NY), let me theorize myself: You shouldn't be using pkt->data_len, it's an internal pkt variable. Also, it's valid only for TX pkt's (it's naming in https://github.com/zephyrproject-rtos/zephyr/commit/753daa65f9181b8ea89a1597b7aaa30a45910f44 is unfortunate, and it should be renamed, actually, I get an idea to make it union with some existing member, which is vice-versa used only for RX pkts.)

pfalcon commented:
So, 2nd commit mesage:

>  net: net_app: set packet data_len on incoming packets

> This fixes a bug when calling net_pkt_append[_all] which uses
pkt->data_len as part of the maximum packet length calculation
when the net_context is set.

The only explanation is that you receive incoming packet, parse it, then try to reuse for sending. Can you please confirm that? And maybe it's not "you" but "net_app code, maybe specifically TLS-related code". Can you confirm that too? I'd say that's a problem, because RX vs TX packets belong to different pools. Even if that works w/o problems, that's somehow undermines the idea of having separate pools (effectively, there're separate pools for benefits they give (e.g. avoid RX-by-TX and vice-versa starvation), but "reuse" approach effectively works it around, killing the benefits it gives (e.g. there still can be starvation, TX will hog on RX buffers). Why do we have separate pools at all then?

pfalcon commented:
I'm +1 on the 1st commit, the same as @pfl.

pfl commented:
pkt->data_len seems to be set now in the right place in the 2nd commit. I hope that means data_len is always properly set up after this for any RX code path. Why it is needed by the TLS code is bit mysterious still.

mike-scott commented:
@pfalcon 
> And given that you're likely on holidays (Merry Xmas/Happy NY), let me theorize myself: You shouldn't be using pkt->data_len, it's an internal pkt variable. Also, it's valid only for TX pkt's (it's naming in 753daa6 is unfortunate, and it should be renamed, actually, I get an idea to make it union with some existing member, which is vice-versa used only for RX pkts.)

Yes, data_len seem to be an internal variable used only in net_pkt.c which is only set in net_pkt_get() only for TX packets.

However, net_pkt_append() which uses data_len is NOT a function used only on TX packets (grep shows it being used all over the place in Zephyr.   Specifically, the net_app layer is using this function to append decrypted RX packet buffer data prior to returning the packet to app code.

So 1 of 3 things needs to happen:
1) [current solution] set pkt->data_len in the nat_app code so that it can use net_pkt_append() instead of generating a hidden error.
2) Add the same "if (slab != &rx_pkts) {" check to net_pkt_append to avoid data_len for RX packets.
3) There is also a static function "net_pkt_append_bytes()" which could be exposed internally to the subsys/net/ip layers and used instead of net_pkt_append() where the buffer length is already known.  This might be a better fit for the net_app use case.  However, we should update API documentation to specifically state that net_app_append*() functions are only to be used on TX packets.  NOTE: This could easily be misused by developers and seems to be a rabbit hole considering the current state of net_pkt and the movement of net_buf structs in the future.

mike-scott commented:
@pfalcon regarding 2nd commit description

> The only explanation is that you receive incoming packet, parse it, then try to reuse for sending. Can you please confirm that?

No, net_app code is receiving an encrypted incoming packet.  It is decrypted and the data is appended to a new RX packet handed off to the app code.

> And maybe it's not "you" but "net_app code, maybe specifically TLS-related code". Can you confirm that too?

I will add a clearer description of what the patch is solving if it's determined that this is the correct patch for net_app code.

> I'd say that's a problem, because RX vs TX packets belong to different pools. Even if that works w/o problems, that's somehow undermines the idea of having separate pools (effectively, there're separate pools for benefits they give (e.g. avoid RX-by-TX and vice-versa starvation), but "reuse" approach effectively works it around, killing the benefits it gives (e.g. there still can be starvation, TX will hog on RX buffers). Why do we have separate pools at all then?

Not a TX packet, this is an RX packet as described above.

pfalcon commented:
> Specifically, the net_app layer is using this function to append decrypted RX packet buffer data prior to returning the packet to app code.

Thanks for the detailed description. I need to look one day how net_app handles TLS, I'm just afraid of what I can see and how much time I'll spend filing new tickets instead of looking into resolving already filed ;-).

Let me approve this.

mike-scott commented:
@pfl did you have any other changes?

pfl commented:
Looks fine to me. We can always re-fix it if needed later on or so...

lpereira reviewed (approved):
LGTM
